src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs

```diff
@@ -161,32 +163,24 @@ public override void Visit(ITypeDefinition parentType, IEnumerable<IFieldDefinit
                     foreach (var genericField in genericTypedFields)
                         newFields.Add(genericField);
 
-                    if (!genericTypedFields.Any())
+                    // For definiteassignment checks the compiler needs to know there is a private field
+                    // that has not be initialized so if there are any we need to add a dummy private
+                    // field to help the compiler do its job and error about uninitialized structs
+                    bool hasRefPrivateField = excludedFields.Any(f => f.Type.IsOrContainsReferenceType());
+
+                    // If at least one of the private fields contains a reference type then we need to
+                    // set this field type to object or reference field to inform the compiler to block
+                    // taking pointers to this struct because the GC will not track updating those references
+                    if (hasRefPrivateField)
+                    {
+                        DummyFieldWriterHelper(parentType, excludedFields, newFields, parentType.PlatformType.SystemObject);
+                    }
+
+                    bool hasValueTypePrivateField = excludedFields.Any(f => f.Type.IsValueType);
```

Comment:
I think this `IsValueType` check is what caused IntPtr/UIntPtr to loose their dummy fields. I think we should instead simply negate the reference type check instead, or simply filter the list down and anything left consider them to be valuetype fields.